Use logRangesFlag in API, route reads based on TreeID

lkatalin opened the pull request:
#### Summary
This PR hooks up the `logRangeMap` flag to be used in API construction. The map can then be accessed to do routing for uuid-based and log-index-based reads.

I made the `logRangeMap` flag into a string because the `Set()` method can be used for construction and it was easier than trying to cram the data into a struct from the flag. But if there's a clean way to do this that doesn't involve this change, I'm happy to do it.

WIP because I still need to add some tests, ~~fix the e2e test~~, and do some cleanup, ex. maybe around the int types. And I need to think through 1) making sure this will not break old clients, 2) do we need to use this logic in the writes too before merging (I think this will actually be a change made to the pending PR that emits longer UUIDs upon artifact upload).

Questions:
- [x] How can the current TreeID be accessed during e2e testing?
  - Instead of this, I added a condition that if the `treeID` is zero, we use the `api.logID` instead 
- [x] In the case that both a `tlog_id` and `logRangeMap` are passed in, which takes precedence? The `logRangeMap`'s `ActiveIndex()` **should be** the `tlog_id` (currently using `ActiveIndex()`). 
  - Should we add a check to make sure these are the same? Edit: not doing this
  - Should we base the `tlog_id` in the API on the `ActiveIndex`? Edit: no, I'm appending the `tlog_id` to the `rangeMap`
  - Or should the `tlog_id` function as a flag for signaling the need for a new shard? This seems to be how it is used currently via `createAndInitTree`. Edit: yes, still doing this

Note: Currently it looks like the `tlog_id` in the API is used in only [one place](https://github.com/sigstore/rekor/blob/adab5c53055aa020e0d70da1ad25a394376a9aa6/pkg/api/api.go#L82) in `api.go`, to check whether it is zero and determine if the program should `createAndInitTree`.

Edit: So I think my answer to the above is to add logic that will append any nonzero `tlog_id` passed in at server startup to the API's `logRange`, see code from this PR [here](https://github.com/sigstore/rekor/blob/25b2ac8c0bb61a9bf059e0fad6b011900ef404c8/pkg/api/api.go#L92). If this isn't the right way, let me know or I guess we will change it later. 

#### Ticket Link
Fixes #670
Fixes #632 
Fixes #629

lkatalin committed:
Update API based on logRangesFlag

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lkatalin committed:
Use API's logRanges to retrieve artifacts

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lukehinds reviewed (approved):
LGTM

bobcallaway commented on cmd/rekor-server/app/serve.go:
```diff
@@ -103,6 +103,14 @@ var serveCmd = &cobra.Command{
+		rangeMap := viper.GetString("trillian_log_server.log_id_ranges")
+		if rangeMap != "" {
+			if err := logRangeMap.Set(rangeMap); err != nil {
+				log.Logger.Errorf("unable to set logRangeMap from flag: %v", err)
```
if we can't set the `logRangeMap` should this error be fatal?

bobcallaway commented on pkg/api/entries.go:
```diff
@@ -117,9 +117,11 @@ func logEntryFromLeaf(ctx context.Context, signer signature.Signer, tc TrillianC
-	tc := NewTrillianClient(ctx)
+	tid, resolvedIndex := api.logRanges.ResolveVirtualIndex(int(params.LogIndex))
+	tc := NewTrillianClientFromTreeID(ctx, int64(tid))
+	log.Logger.Debugf("Retrieving resolved index %v from TreeID %v", resolvedIndex, tid)
```
```suggestion
	log.RequestIDLogger(params.HTTPRequest).Debugf("Retrieving resolved index %v from TreeID %v", resolvedIndex, tid)
```

bobcallaway commented on pkg/api/entries.go:
```diff
@@ -272,8 +274,23 @@ func GetLogEntryByUUIDHandler(params entries.GetLogEntryByUUIDParams) middleware
-	tc := NewTrillianClient(params.HTTPRequest.Context())
+
+	tc := NewTrillianClientFromTreeID(params.HTTPRequest.Context(), tid)
+	log.Logger.Debugf("Retrieving UUID %v from TreeID %v", entryUUID, tid)
```
```suggestion
	tc := NewTrillianClientFromTreeID(ctx, tid)
	log.RequestIDLogger(params.HTTPRequest).Debugf("Retrieving UUID %v from TreeID %v", entryUUID, tid)
```

bobcallaway commented on pkg/sharding/sharding.go:
```diff
@@ -124,3 +124,27 @@ func GetUUIDFromIDString(id string) (string, error) {
+// Returns TreeID from an EntryID string
+// TODO: Add some tests
+func GetTreeIDFromEntryIDString(id string) (int64, error) {
+	if len(id) == UUIDHexStringLen {
```
style: rewrite this as a switch statement on `len(id)`?

lkatalin committed:
Add helpers and refactor sharding package

- Adds a function to get a TreeID from an ID string
- Adds testing for the above
- Consolidates validation logic for UUID, TreeID, EntryID
- Removes code that attempts to use ActiveIndex() in the sharding
  package, as this is not accessible due to import cycles
- Other small cleanup and typo fixes

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lkatalin committed:
Change logRanges to work with int64

This is the type used by the trillian TreeID and saves from having to
convert in multiple places.

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

lkatalin committed:
Add TreeID to LogInfo API endpoint

WARNING: breaks loginfo cmd to current prod server

Signed-off-by: Lily Sturmann <lsturman@redhat.com>
